[test]: Update test workflow to improve test coverage metrics

hickeyma opened the pull request:
This PR is based on a discussion with @YaoJiayi where use case ask is: "some coverage metrics for unit tests".  There are metrics generated currently using [coverage](https://coverage.readthedocs.io/en/latest/) and this PR is to improve diplay and availabiity of the metrics.

Replacing [coverage](https://coverage.readthedocs.io/en/latest/) with pytest plugins [pytest-cov](https://pytest-cov.readthedocs.io/en/latest/) and [pytest-html](https://pytest-html.readthedocs.io/en/latest/) which provide more functionality and consistency than `coverage`.

The following is now generated when tests are run:

- HTML and XML report of test results
- Summary of test coverage to terminal


The terminal output is as follows:

```shell
[...]

===================================================================== test session starts ======================================================================
platform linux -- Python 3.10.12, pytest-8.4.1, pluggy-1.6.0
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /var/lib/buildkite-agent/builds/unit-end-to-end-test-1/lmcache/lmcache-unittests
configfile: pyproject.toml
plugins: buildkite-test-collector-1.0.4, html-4.1.1, metadata-3.1.1, benchmark-5.1.0, cov-6.2.1, anyio-4.9.0
collected 235 items                                                                                                                                            
tests/benchmarks/test_benchmark.py ......                                                                                                                [  2%]
tests/test_backends.py ...........................                                                                                                       [ 14%]
tests/test_blend.py ......                                                                                                                               [ 16%]
tests/test_cache_engine.py ......................................................                                                                        [ 39%]
tests/test_connector.py ...                                                                                                                              [ 40%]
tests/test_evictor.py .....                                                                                                                              [ 42%]
tests/test_mempool.py ......                                                                                                                             [ 45%]
tests/test_observability.py ......                                                                                                                       [ 48%]
tests/test_protocol.py ..                                                                                                                                [ 48%]
tests/test_serde.py ...........                                                                                                                          [ 53%]
tests/v1/test_cache_engine.py ................................                                                                                           [ 67%]
tests/v1/test_config.py ..                                                                                                                               [ 68%]
tests/v1/test_connector.py ..........                                                                                                                    [ 72%]
tests/v1/test_gds.py ..                                                                                                                                  [ 73%]
tests/v1/test_gpu_connector.py .......s.                                                                                                                 [ 77%]
tests/v1/test_mem_kernels.py ....................                                                                                                        [ 85%]
tests/v1/test_memory_management.py ..................                                                                                                    [ 93%]
tests/v1/test_remote_mla_worker_id_as0.py .                                                                                                              [ 93%]
tests/v1/test_token_database.py ..............                                                                                                           [ 99%]
tests/v1/test_weka.py .                                                                                                                                  [100%]
======================================================================= warnings summary =======================================================================
lmcache/usage_context.py:26
  /var/lib/buildkite-agent/builds/unit-end-to-end-test-1/lmcache/lmcache-unittests/lmcache/usage_context.py:26: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
    import pkg_resources
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================== tests coverage ========================================================================
_______________________________________________________ coverage: platform linux, python 3.10.12-final-0 _______________________________________________________
Name                                                                 Stmts   Miss  Cover
----------------------------------------------------------------------------------------
lmcache/__init__.py                                                      0      0   100%
lmcache/_version.py                                                     13     13     0%
lmcache/blend/__init__.py                                                0      0   100%
lmcache/blend/executor.py                                               88     15    83%
lmcache/blend/interfaces.py                                             30      3    90%
lmcache/blend/retriever.py                                             113     13    88%
lmcache/cache_engine.py                                                220     22    90%
lmcache/config.py                                                      127      5    96%
lmcache/connections.py                                                  74     48    35%
lmcache/integration/__init__.py                                          0      0   100%
lmcache/integration/vllm/__init__.py                                     0      0   100%
lmcache/integration/vllm/lmcache_connector_v1.py                        35     35     0%
lmcache/integration/vllm/utils.py                                       22     13    41%
lmcache/integration/vllm/vllm_adapter.py                               469    469     0%
lmcache/integration/vllm/vllm_v1_adapter.py                            357    357     0%
lmcache/logging.py                                                      39      6    85%
lmcache/observability.py                                               279      7    97%
lmcache/protocol.py                                                     40      0   100%
lmcache/server/__init__.py                                               0      0   100%
lmcache/server/__main__.py                                              79     79     0%
lmcache/server/server_storage_backend/__init__.py                       12     12     0%
lmcache/server/server_storage_backend/abstract_backend.py               21     21     0%
lmcache/server/server_storage_backend/local_backend.py                  94     94     0%
lmcache/storage_backend/__init__.py                                     29      0   100%
lmcache/storage_backend/abstract_backend.py                             38      6    84%
lmcache/storage_backend/connector/__init__.py                           55      7    87%
lmcache/storage_backend/connector/base_connector.py                     73     15    79%
lmcache/storage_backend/connector/lm_connector.py                       58      4    93%
lmcache/storage_backend/connector/redis_connector.py                    67      4    94%
lmcache/storage_backend/evictor/__init__.py                              3      0   100%
lmcache/storage_backend/evictor/base_evictor.py                         37      6    84%
lmcache/storage_backend/evictor/lru_evictor.py                          27      2    93%
lmcache/storage_backend/hybrid_backend.py                               67     22    67%
lmcache/storage_backend/local_backend.py                               253     32    87%
lmcache/storage_backend/mem_pool/__init__.py                             3      0   100%
lmcache/storage_backend/mem_pool/base_pool.py                           16      2    88%
lmcache/storage_backend/mem_pool/local_pool.py                          58      5    91%
lmcache/storage_backend/remote_backend.py                              143     27    81%
lmcache/storage_backend/serde/__init__.py                               24      4    83%
lmcache/storage_backend/serde/cachegen_basics.py                        96     26    73%
lmcache/storage_backend/serde/cachegen_decoder.py                       94     13    86%
lmcache/storage_backend/serde/cachegen_encoder.py                      168     70    58%
lmcache/storage_backend/serde/fast_serde.py                             18      7    61%
lmcache/storage_backend/serde/safe_serde.py                             20      0   100%
lmcache/storage_backend/serde/serde.py                                  35      2    94%
lmcache/storage_backend/serde/torch_serde.py                            20      0   100%
lmcache/usage_context.py                                               168    128    24%
lmcache/utils.py                                                        92     23    75%
lmcache/v1/__init__.py                                                   0      0   100%
lmcache/v1/api_server/__init__.py                                        0      0   100%
lmcache/v1/api_server/__main__.py                                      150    150     0%
lmcache/v1/cache_controller/__init__.py                                  3      0   100%
lmcache/v1/cache_controller/controller_manager.py                       68     68     0%
lmcache/v1/cache_controller/controllers/__init__.py                      3      3     0%
lmcache/v1/cache_controller/controllers/kv_controller.py                60     60     0%
lmcache/v1/cache_controller/controllers/registration_controller.py      61     61     0%
lmcache/v1/cache_controller/executor.py                                 58     41    29%
lmcache/v1/cache_controller/message.py                                 182     39    79%
lmcache/v1/cache_controller/rpc_utils.py                                31     22    29%
lmcache/v1/cache_controller/worker.py                                   97     75    23%
lmcache/v1/cache_engine.py                                             321    132    59%
lmcache/v1/cache_interface.py                                            5      5     0%
lmcache/v1/compute/__init__.py                                           0      0   100%
lmcache/v1/compute/attention/__init__.py                                 0      0   100%
lmcache/v1/compute/attention/abstract.py                                 7      7     0%
lmcache/v1/compute/attention/flash_attn.py                              25     25     0%
lmcache/v1/compute/attention/metadata.py                                12     12     0%
lmcache/v1/compute/blend/__init__.py                                     2      0   100%
lmcache/v1/compute/blend/blender.py                                     60     49    18%
lmcache/v1/compute/blend/metadata.py                                    17      3    82%
lmcache/v1/compute/blend/utils.py                                       25     12    52%
lmcache/v1/compute/models/__init__.py                                    0      0   100%
lmcache/v1/compute/models/llama.py                                      53     53     0%
lmcache/v1/compute/models/utils.py                                      23     12    48%
lmcache/v1/compute/positional_encoding.py                               91     91     0%
lmcache/v1/config.py                                                   255     34    87%
lmcache/v1/distributed_server/__init__.py                                3      0   100%
lmcache/v1/distributed_server/abstract_server.py                        17      4    76%
lmcache/v1/distributed_server/naive_server.py                          110     86    22%
lmcache/v1/gpu_connector.py                                            473    205    57%
lmcache/v1/lookup_server/__init__.py                                     3      0   100%
lmcache/v1/lookup_server/abstract_server.py                             19      5    74%
lmcache/v1/lookup_server/redis_server.py                                44     29    34%
lmcache/v1/memory_management.py                                        569     90    84%
lmcache/v1/protocol.py                                                  73      8    89%
lmcache/v1/server/__init__.py                                            0      0   100%
lmcache/v1/server/__main__.py                                           84     84     0%
lmcache/v1/server/storage_backend/__init__.py                           11     11     0%
lmcache/v1/server/storage_backend/abstract_backend.py                   23     23     0%
lmcache/v1/server/storage_backend/local_backend.py                      31     31     0%
lmcache/v1/server/utils.py                                              11     11     0%
lmcache/v1/storage_backend/__init__.py                                  45      5    89%
lmcache/v1/storage_backend/abstract_backend.py                          40     11    72%
lmcache/v1/storage_backend/connector/__init__.py                       185     36    81%
lmcache/v1/storage_backend/connector/audit_connector.py                 74     57    23%
lmcache/v1/storage_backend/connector/base_connector.py                  22      5    77%
lmcache/v1/storage_backend/connector/blackhole_connector.py             20      6    70%
lmcache/v1/storage_backend/connector/fs_connector.py                    69     14    80%
lmcache/v1/storage_backend/connector/infinistore_connector.py           91     65    29%
lmcache/v1/storage_backend/connector/instrumented_connector.py          40      1    98%
lmcache/v1/storage_backend/connector/lm_connector.py                    64      4    94%
lmcache/v1/storage_backend/connector/mooncakestore_connector.py        122     81    34%
lmcache/v1/storage_backend/connector/nixl_connector.py                 258    258     0%
lmcache/v1/storage_backend/connector/nixl_connector_v2.py              374    374     0%
lmcache/v1/storage_backend/connector/nixl_utils.py                      38     38     0%
lmcache/v1/storage_backend/connector/redis_connector.py                133     29    78%
lmcache/v1/storage_backend/evictor/__init__.py                           3      0   100%
lmcache/v1/storage_backend/evictor/base_evictor.py                      17      2    88%
lmcache/v1/storage_backend/evictor/lru_evictor.py                       30     10    67%
lmcache/v1/storage_backend/gds_backend.py                              357    122    66%
lmcache/v1/storage_backend/local_cpu_backend.py                        224    104    54%
lmcache/v1/storage_backend/local_disk_backend.py                       190     56    71%
lmcache/v1/storage_backend/naive_serde/__init__.py                      20      2    90%
lmcache/v1/storage_backend/naive_serde/cachegen_basics.py               35     13    63%
lmcache/v1/storage_backend/naive_serde/cachegen_decoder.py              61      4    93%
lmcache/v1/storage_backend/naive_serde/cachegen_encoder.py              44      2    95%
lmcache/v1/storage_backend/naive_serde/kivi_serde.py                    12      4    67%
lmcache/v1/storage_backend/naive_serde/naive_serde.py                   11      0   100%
lmcache/v1/storage_backend/naive_serde/serde.py                         11      2    82%
lmcache/v1/storage_backend/nixl_backend.py                             175    175     0%
lmcache/v1/storage_backend/remote_backend.py                           155     33    79%
lmcache/v1/storage_backend/storage_manager.py                          161     62    61%
lmcache/v1/storage_backend/weka_gds_backend.py                         264     77    71%
lmcache/v1/token_database.py                                           101      9    91%
----------------------------------------------------------------------------------------
TOTAL                                                                10300   4819    53%
Coverage HTML written to dir coverage-test
Coverage XML written to file coverage-test.xml
------------------------------------------------------------------------------------------------------------ benchmark: 7 tests ------------------------------------------------------------------------------------------------------------
Name (time in us)                                        Min                    Max                   Mean                StdDev                 Median                   IQR            Outliers          OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_vllm_paged_connector_v2_to_gpu_bench            79.5568 (1.0)          90.3820 (1.0)          80.7371 (1.0)          1.6905 (1.0)          80.2190 (1.0)          0.5042 (1.0)         10;14  12,385.8826 (1.0)         100        1000
test_cachegen_decoder_bench[64-huggingface]       6,024.5739 (75.73)     7,617.0880 (84.28)     6,476.8351 (80.22)      348.9517 (206.42)    6,430.1135 (80.16)      539.6837 (>1000.0)      33;2     154.3964 (0.01)        110           1
test_cachegen_decoder_bench[64-vllm]              6,610.3823 (83.09)    13,813.6926 (152.84)   12,813.2847 (158.70)   1,344.8095 (795.53)   13,114.9660 (163.49)     365.2431 (724.39)        2;4      78.0440 (0.01)         45           1
test_cachegen_decoder_bench[256-vllm]             9,490.5831 (119.29)   12,092.8800 (133.80)    9,967.5643 (123.46)     407.0629 (240.80)    9,918.0075 (123.64)     338.6994 (671.75)        6;1     100.3254 (0.01)         44           1
test_cachegen_decoder_bench[256-huggingface]      9,607.6690 (120.76)   13,601.4670 (150.49)   10,977.5291 (135.97)     974.6909 (576.59)   10,866.7510 (135.46)   1,235.1756 (>1000.0)      15;2      91.0952 (0.01)         45           1
test_cachegen_decoder_bench[768-huggingface]     27,605.6598 (346.99)   33,811.5389 (374.10)   28,820.8794 (356.97)   1,292.7517 (764.74)   28,485.0169 (355.09)     774.6944 (>1000.0)       1;1      34.6971 (0.00)         21           1
test_cachegen_decoder_bench[768-vllm]            27,619.5561 (347.17)   31,759.7990 (351.40)   28,781.6715 (356.49)   1,058.2254 (626.00)   28,785.3014 (358.83)   1,229.9146 (>1000.0)       6;1      34.7443 (0.00)         23           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
-------------- Generated html report: file:///var/lib/buildkite-agent/builds/unit-end-to-end-test-1/lmcache/lmcache-unittests/durations/test.html --------------
==================================================== 234 passed, 1 skipped, 1 warning in 626.74s (0:10:26) =====================================================
buildkite-test-collector - WARNING - No `BUILDKITE_ANALYTICS_TOKEN` environment variable present

[...]
```

hickeyma committed:
Update test workflow to use pytest plugins instead of coverage

Pytest plugins provide more functionality and consistency than
`coverage`

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

Shaoting-Feng commented on .buildkite/install-env-bm.sh:
```diff
@@ -22,7 +22,10 @@ set -xe
-pip install coverage
+
+# Added these install as latestrequirements/test.txt is not being installed
+uv pip install pytest-cov
```
`install-env-bm.sh` and `install-env-e2e.sh` are actually outdated and no longer used in buildkite.

hickeyma replied:
Ok, removed changes from the file for now. I'll remove the files all together in a separate commit once I get the tests running.

hickeyma replied:
Files removed in 4628a1d3f427c4fec6667942e82935130ce5ac4b

Shaoting-Feng commented on .buildkite/install-env.sh:
```diff
@@ -18,7 +18,10 @@ CUDA_VERSION="12.1"
-uv pip install coverage
+
+# Added these install as latestrequirements/test.txt is not being installed
+uv pip install pytest-cov
```
In `.buildkite/pipeline.yml` line 24, the restored venv of step 2 might hit the venv of step 1 from previous run, because `requirements/common.txt` and `requirements/test.txt` in the key didn't change. Then the venv does not include `pytest-cov` and `pytest-html`.

hickeyma replied:
Ok. I investigated this but can't find how to regenerate the virtual environment so that it will run the install of the updated test requirement file.

@Shaoting-Feng How do I trigger this?

Shaoting-Feng replied:
I suggest try to include the unique build ID (BUILDKITE_BUILD_ID) in the cache key so that each build has its own key. Then only the .venv saved in the current build is restored.

hickeyma replied:
Can you provide more details to explain this?

Shaoting-Feng replied:
Like this:
```yaml
plugins:
      - cache#v1.5.2:
          key: "venv-{{ BUILDKITE_BUILD_ID }}-{{ checksum \"requirements/common.txt\" }}-{{ checksum \"requirements/test.txt\" }}"
          path: ".venv"
          save: "pipeline"
```

hickeyma replied:
Thanks @Shaoting-Feng. I tried your suggestion above with the build id but doesn't seem to be working. Any ideas?

Shaoting-Feng replied:
The plugin only allows a single cache save per pipeline, regardless of the cache key. I’ve added force: true (https://github.com/buildkite-plugins/cache-buildkite-plugin?tab=readme-ov-file#force-boolean-save-only) to force it to overwrite and save on each run.

hickeyma replied:
Thank you @Shaoting-Feng for fixing the caching issue and enabling the test pipeline to pickup the latest test requirements file for pytest coverage dependencies. Pipeline passing now.

hickeyma committed:
Add unique build id to regenerate buildkite env

Add unique build ID (BUILDKITE_BUILD_ID) in the cache key so that each build has its own key.
This is to regenerate the virtual environment so that it will run the install of the updated
test requirement file.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

Commit:
Test packages in step 1

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

Commit:
Test packages in the initial stage of step 2

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

Commit:
Debug buildkite id

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

Commit:
Use key-extra instead of key

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

Commit:
Double check

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

Commit:
Force save

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

hickeyma committed:
Remove temporary check

Rollback commit 6ba45169a14c73a2fcc3d942066af252094a8ab9

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Make HTML coverage file available after test run

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Remove redundant files

Feedback from @Shaoting in https://github.com/LMCache/LMCache/pull/823#discussion_r2143652659
that files are no longr used.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@Shaoting-Feng @YaoJiayi This is updated and ready for review again.

YaoJiayi commented:
> @Shaoting-Feng @YaoJiayi This is updated and ready for review again.

Are the results available on github too? It'll be seen in the buildkite too, right?

YaoJiayi reviewed (approved):
LGTM!

hickeyma commented:
> Are the results available on github too? It'll be seen in the buildkite too, right?

@YaoJiayi I pushed #944 as a follow on to this PR which was merged before I was able to respond.
